src/cpp/rtps/reader/BaseReader.cpp

```diff
@@ -0,0 +1,450 @@
+        auto spourious_record = history_state_->history_record.find(guid);
+        if (spourious_record != history_state_->history_record.end())
+        {
+            EPROSIMA_LOG_INFO(RTPS_READER, "Sporious record found, changing guid "
+                    << guid << " for persistence guid " << persistence_guid);
+            update_last_notified(guid, spourious_record->second);
+            history_state_->history_record.erase(spourious_record);
+        }
+    }
+}
+
+void BaseReader::remove_persistence_guid(
+        const fastrtps::rtps::GUID_t& guid,
+        const fastrtps::rtps::GUID_t& persistence_guid,
+        bool removed_by_lease)
+{
+    std::lock_guard<decltype(mp_mutex)> guard(mp_mutex);
+    auto persistence_guid_stored = (fastrtps::rtps::c_Guid_Unknown == persistence_guid) ? guid : persistence_guid;
+    history_state_->persistence_guid_map.erase(guid);
+    auto count = --history_state_->persistence_guid_count[persistence_guid_stored];
+    if (count <= 0 && may_remove_history_record(removed_by_lease))
+    {
+        history_state_->history_record.erase(persistence_guid_stored);
+        history_state_->persistence_guid_count.erase(persistence_guid_stored);
+    }
+}
+
+fastrtps::rtps::SequenceNumber_t BaseReader::get_last_notified(
+        const fastrtps::rtps::GUID_t& guid)
+{
+    fastrtps::rtps::SequenceNumber_t ret_val;
+    std::lock_guard<decltype(mp_mutex)> guard(mp_mutex);
+    fastrtps::rtps::GUID_t guid_to_look = guid;
+    auto p_guid = history_state_->persistence_guid_map.find(guid);
+    if (p_guid != history_state_->persistence_guid_map.end())
+    {
+        guid_to_look = p_guid->second;
+    }
+
+    auto p_seq = history_state_->history_record.find(guid_to_look);
+    if (p_seq != history_state_->history_record.end())
+    {
+        ret_val = p_seq->second;
+    }
+
+    return ret_val;
+}
+
+fastrtps::rtps::SequenceNumber_t BaseReader::update_last_notified(
+        const fastrtps::rtps::GUID_t& guid,
+        const fastrtps::rtps::SequenceNumber_t& seq)
+{
+    fastrtps::rtps::SequenceNumber_t ret_val;
+    std::lock_guard<decltype(mp_mutex)> guard(mp_mutex);
+    fastrtps::rtps::GUID_t guid_to_look = guid;
+    auto p_guid = history_state_->persistence_guid_map.find(guid);
+    if (p_guid != history_state_->persistence_guid_map.end())
+    {
+        guid_to_look = p_guid->second;
+    }
+
+    auto p_seq = history_state_->history_record.find(guid_to_look);
+    if (p_seq != history_state_->history_record.end())
+    {
+        ret_val = p_seq->second;
+    }
+
+    if (ret_val < seq)
+    {
+        set_last_notified(guid_to_look, seq);
+        new_notification_cv_.notify_all();
+    }
+
+    return ret_val;
+}
+
+void BaseReader::set_last_notified(
+        const fastrtps::rtps::GUID_t& peristence_guid,
+        const fastrtps::rtps::SequenceNumber_t& seq)
+{
+    history_state_->history_record[peristence_guid] = seq;
```

Comment:
We should protect this

Reply:
Changed into `persist_last_notified_nts`.